homeassistant/components/teslemetry/update.py

```diff
@@ -0,0 +1,104 @@
+"""Update platform for Teslemetry integration."""
+
+from __future__ import annotations
+
+from typing import Any, cast
+
+from tesla_fleet_api.const import Scope
+
+from homeassistant.components.update import UpdateEntity, UpdateEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+
+from .entity import TeslemetryVehicleEntity
+from .models import TeslemetryVehicleData
+
+AVAILABLE = "available"
+DOWNLOADING = "downloading"
+INSTALLING = "installing"
+WIFI_WAIT = "downloading_wifi_wait"
+SCHEDULED = "scheduled"
+
+
+async def async_setup_entry(
+    hass: HomeAssistant, entry: ConfigEntry, async_add_entities: AddEntitiesCallback
+) -> None:
+    """Set up the Teslemetry update platform from a config entry."""
+
+    async_add_entities(
+        TeslemetryUpdateEntity(vehicle, entry.runtime_data.scopes)
+        for vehicle in entry.runtime_data.vehicles
+    )
+
+
+class TeslemetryUpdateEntity(TeslemetryVehicleEntity, UpdateEntity):
+    """Teslemetry Updates entity."""
+
+    def __init__(
+        self,
+        data: TeslemetryVehicleData,
+        scopes: list[Scope],
+    ) -> None:
+        """Initialize the Update."""
+        self.scoped = Scope.VEHICLE_CMDS in scopes
+        super().__init__(
+            data,
+            "vehicle_state_software_update_status",
+        )
+
+    def _async_update_attrs(self) -> None:
+        """Update the attributes of the entity."""
+
+        # Supported Features
+        if self.scoped and self._value in (
+            AVAILABLE,
+            SCHEDULED,
+        ):
+            # Only allow install when an update is available
+            self._attr_supported_features = (
+                UpdateEntityFeature.PROGRESS | UpdateEntityFeature.INSTALL
+            )
+        else:
+            self._attr_supported_features = UpdateEntityFeature.PROGRESS
```

Comment:
Euh, isn't the supported features to tell what features the entity has at all times? E.g. if its not showing progress it doesn't mean the progress flag should be removed.

Reply:
You're correct about progress, but the install capability needs to be removed until an upgrade can be installed. There was a change recently that allows capabilities to change dynamically, I'll see if I can find the reference.

Reply:
Why should the install be removed then? It should only be removed if there is an update that can't be installed iirc

Reply:
As discussed, for the records, because it cant be installed why downloading, which can be common if the vehicle doesnt have WiFi were its parked. This was also discussed for Tessie as is identical.